compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -509,7 +509,9 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                       }) =>
                 {
                   // FIXME(associated_const_equality) when does this unwrap fail? I have no idea what case it would.
-                  let trait_def_id = trait_ref.trait_def_id().unwrap();
+                  let Some(trait_def_id) = trait_ref.trait_def_id() else {
```

Comment:
The comment above this line can be removed or change to explain the situation where this arises.